genesis/engine/entities/fem_entity.py

```diff
@@ -744,12 +745,134 @@ def set_muscle_direction(self, muscle_direction):
+            link_init_quat,
+            envs_idx
+        )
+
+    def update_constraint_targets(self, verts_idx, target_poss, envs_idx=None):
+        """Update target positions for existing constraints."""
+        if not self._solver._constraints_initialized:
+            gs.logger.warning("Ignoring update_constraint_targets; constraints have not been initialized.")
+            return
+
+        envs_idx = self._scene._sanitize_envs_idx(envs_idx)
+        verts_idx = self._sanitize_input_to_batch_tensor(verts_idx, dtype=gs.tc_int) + self._v_start
+        target_poss = self._sanitize_input_to_batch_tensor(target_poss, gs.tc_float, dims=2)
+
+        assert target_poss.shape[1] == verts_idx.shape[1], "Target position should be provided for each vertex."
+        self._assert_verts_idx_in_bounds(verts_idx)
+
+        self._solver._kernel_update_constraint_targets(verts_idx, target_poss, envs_idx)
+
+    def remove_vertex_constraints(self, verts_idx=None, envs_idx=None):
+        """Remove constraints from specified vertices, or all if None."""
+        if not self._solver._constraints_initialized:
+            gs.logger.warning("Ignoring remove_vertex_constraints; constraints have not been initialized.")
+            return
+
+        if verts_idx is None:
+            self._solver.vertex_constraints.is_constrained.fill(0)
+        else:
+            verts_idx = self._sanitize_input_to_batch_tensor(verts_idx, dtype=gs.tc_int) + self._v_start
+            self._assert_verts_idx_in_bounds(verts_idx)
+            envs_idx = self._scene._sanitize_envs_idx(envs_idx)
+            self._solver._kernel_remove_specific_constraints(verts_idx, envs_idx)
+
+    @ti.kernel
+    def _kernel_get_verts_pos(self, f: ti.i32, pos: ti.types.ndarray(), verts_idx: ti.types.ndarray()):
+        # get current position of vertices
+        n = verts_idx.shape[0]
+        B = verts_idx.shape[1]
+        for i_v, i_b in ti.ndrange(n, B):
+            i_global = verts_idx[i_v, i_b] + self.v_start
+            for j in ti.static(range(3)):
+                pos[i_b, i_v, j] = self._solver.elements_v[f, i_global, i_b].pos[j]
+
+    def _sanitize_input_to_batch_tensor(self, arr, dtype, dims=1):
+        if not isinstance(arr, torch.Tensor):
+            arr = torch.as_tensor(arr, dtype=dtype, device=gs.device)
+        tensor = torch.atleast_1d(arr)
+        if tensor.ndim < dims+1:
+            # add batch dimension if missing
+            tensor = tensor.unsqueeze(0).repeat((self._sim._B, *([1] * dims)))
+            # tensor = tensor.unsqueeze(0).repeat(self._sim._B, 1)
+        assert tensor.shape[0] == self._sim._B, "Input tensor batch size must match the number of environments."
+        return tensor
+
+    def _assert_verts_idx_in_bounds(self, verts_idx):
+        """
+        Assert that all vertex indices are within the valid range.
+        """
+        out_of_bounds = torch.any(torch.logical_or(verts_idx >= self._solver.n_vertices, verts_idx < 0))
```

Comment:
`out_of_bounds = not (0 <= verts_idx & verts_idx < self._solver.n_vertices).all()`